src/GraphQL/Federation/Extensions/FederationMetadataExtensions.cs

```diff
@@ -0,0 +1,85 @@
+using GraphQL.Types;
+using static GraphQL.Federation.FederationHelper;
+
+namespace GraphQL.Federation;
+
+/// <summary>
+/// Extension methods to configure federation directives within GraphQL schemas.
+/// These methods allow the application of various federation directives such as @key, @shareable, @inaccessible, @override, @provides, and @requires to GraphQL types and fields.
+/// </summary>
+public static class FederationMetadataExtensions
+{
+    /// <summary>
+    /// Adds the "@key" directive to a GraphQL type.
+    /// </summary>
+    /// <param name="graphType">The GraphQL type to which the directive is added.</param>
+    /// <param name="fields">An array of field names that form the key.</param>
+    /// <param name="resolvable">Indicates whether the key is resolvable. Default is true.</param>
+    /// <typeparam name="TMetadataWriter">The type of the metadata writer.</typeparam>
+    /// <returns>The modified metadata writer.</returns>
+    /// <remarks>
+    /// See <see href="https://www.apollographql.com/docs/federation/federated-types/federated-directives#key"/>.
+    /// </remarks>
+    public static TMetadataWriter Key<TMetadataWriter>(this TMetadataWriter graphType, string[] fields, bool resolvable = true)
```

Comment:
This extension is useful to define compound or nested field keys. But I guess a single field as a key should be the most frequently used. Then need another extension based on `IFieldMetadataWriter`

Reply:
Actually, FieldBuilder implements `IMetadataWriter`, so it works for all cases.

Reply:
But it doesn't implement `IComplexGraphType` so it doesn't work...

Reply:
I don't follow you. `@key` is only valid on an object or interface, which is what `IComplexGraphType` is.

```gql
directive @key(fields: FieldSet!, resolvable: Boolean = true) repeatable on OBJECT | INTERFACE
```

https://www.apollographql.com/docs/federation/federated-types/federated-directives#key

Reply:
As is implemented now, these methods simply add directives to the schema.  They are not 'intelligent'; for example, you can't put `[Key]` on a field and have it automatically add the proper directive to the graph type.  While that's possible, it's not the current design.

So for `@key`, you can do `@key(fields: "id name pricing { id }")` by specifying `this.Key("id name pricing { id }")`

Reply:
Note that trying to make `@key` 'intelligent' presents a problem where the field case is not determined until the schema is initialized.  So when you write `Field(x => x.Id)` that will initially have an name of "Id" which would probably not be valid as an argument to `@key`.  It would need to be "id".  But for schemas using pascal case, it would need to be "Id".  So I decided to ignore that set of problems entirely.  And that ignores trying to handle complex keys with child properties.

Reply:
I mean, probably, `@key(fields: "id")` is the most commonly used. Then I'd expect to be able to do this
```
Field(x => x.Id).Key();
```
And the field name will be taken from the FieldBuilder's `Name`.

Reply:
Note that technically `IInputObjectGraphType` inherits `IComplexGraphType` also.  I'm not too concerned about it; schema validation will ensure all directives are placed appropriately.  We could split the method, as I did with some of the other extension methods, having one target `IObjectGraphType` and the other target `IInterfaceGraphType`.  I would need to add another class.

Thoughts?

Reply:
> It would need to be "id"

I believe people will be using `nameof()` to get the field name anyway, so this problem won't go anywhere

Reply:
Well, it would be nice. Maybe we could add a schema visitor to build the keys after the schema is initialized.

Reply:
>having one target IObjectGraphType and the other target IInterfaceGraphType

Sounds better than implementing an analyzer to prevent incorrect usage. We can probably use the `[AllowedOn<>]`, but I'd prefer strong API over static analysis.

Reply:
In your example `Field(x => x.Id).Key();` doesn’t even have access to the graph type, so it would have to be something like a schema visitor.

Reply:
I see. Then, it can be enchantment in 8.1 if you prefer.

Reply:
There is still plenty of space for analyzers. For example, validating field names really exists. Or key field doesn't return union, interface or takes arguments.

Reply:
> I see. Then, it can be enchantment in 8.1 if you prefer.

Yes let’s just get a workable version pushed out

Reply:
> > having one target IObjectGraphType and the other target IInterfaceGraphType
> 
> Sounds better than implementing an analyzer to prevent incorrect usage. We can probably use the `[AllowedOn<>]`, but I'd prefer strongly typed API over static analysis.

I’ll make the change

Reply:
> > I see. Then, it can be enchantment in 8.1 if you prefer.
> 
> Yes let’s just get a workable version pushed out

Or at least merged in so we can review what further enhancements we want to make

Reply:
Btw key on interface is only supported sine v2.3. Adding it now will require adding version detection/configuration + validation.

Reply:
Yeah, we need to work on that (not yet in #3921 )

Reply:
Generally, I'm planning to have the proper directives added based on the version.  Schema validation will already match up the applied directives to configured directives.  So it will catch this error if it's the wrong version.

Reply:
> > > having one target IObjectGraphType and the other target IInterfaceGraphType
> > 
> > 
> > Sounds better than implementing an analyzer to prevent incorrect usage. We can probably use the `[AllowedOn<>]`, but I'd prefer strongly typed API over static analysis.
> 
> I’ll make the change

Done